I did see it fail in 

```
=== Failed
=== FAIL: amd64.integration.daemon TestLiveRestore/volume_references/local_volume_with_mount_options (2.34s)
    daemon_test.go:581: assertion failed: error is not nil: Error response from daemon: remove test-live-restore-volume-references-local: volume has active mounts
        --- FAIL: TestLiveRestore/volume_references/local_volume_with_mount_options (2.34s)

=== FAIL: amd64.integration.daemon TestLiveRestore/volume_references (16.91s)
    --- FAIL: TestLiveRestore/volume_references (16.91s)

=== FAIL: amd64.integration.daemon TestLiveRestore (0.00s)
```

but only in [test (graphdriver) / integration (ubuntu-20.04, rootless)](https://github.com/moby/moby/actions/runs/7570683298/job/20616570235?pr=47105#logs).

I have no idea if that's a flake, or happens to be using the filesystem that needs to not be ref-counted, and triggers it because of how it's using mount options. Different mount options seem like something that should produce different mounts, and I don't see any code in `func (m *refCountMounter) Mount(mounts []mount.Mount, containerID string) (target string, retErr error)` to ensure we don't accidentally inc-ref when passed read-write `mounts` when previously called with read-only `mounts`. Maybe that's not possible by usage, but the API appears to allow it.

--- a/daemon/snapshotter/mount.go
+++ b/daemon/snapshotter/mount.go
@@ -13,9 +13,6 @@ import (
 	"github.com/moby/sys/mountinfo"
 )
 
-// List of known filesystems that can't be re-mounted or have shared layers
-var refCountedFileSystems = []string{"fuse-overlayfs", "overlayfs", "stargz", "zfs"}
-
 // Mounter handles mounting/unmounting things coming in from a snapshotter
 // with optional reference counting if needed by the filesystem
 type Mounter interface {
@@ -27,34 +24,17 @@ type Mounter interface {
 	Mounted(containerID string) (string, error)
 }
 
-// inSlice tests whether a string is contained in a slice of strings or not.
-// Comparison is case sensitive
-func inSlice(slice []string, s string) bool {
-	for _, ss := range slice {
-		if s == ss {
-			return true
-		}
-	}
-	return false
-}
-
 // NewMounter creates a new mounter for the provided snapshotter
-func NewMounter(home string, snapshotter string, idMap idtools.IdentityMapping) Mounter {
-	mnter := mounter{
-		home:        home,
-		snapshotter: snapshotter,
-		idMap:       idMap,
+func NewMounter(home string, snapshotter string, idMap idtools.IdentityMapping) *refCountMounter {
+	return &refCountMounter{
+		base: mounter{
+			home:        home,
+			snapshotter: snapshotter,
+			idMap:       idMap,
+		},
+		rc:     graphdriver.NewRefCounter(checker()),
+		locker: locker.New(),
 	}
-
-	if inSlice(refCountedFileSystems, snapshotter) {
-		return &refCountMounter{
-			base:   mnter,
-			rc:     graphdriver.NewRefCounter(checker()),
-			locker: locker.New(),
-		}
-	}
-
-	return &mnter
 }
 
 type refCountMounter struct {

